### ISSUE TITLE
Node runs out of memory running our builds

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#5229]
I've seen it crash on JS bundle tasks and `build/pure-packages.dart`. @hterkelsen reports it crashing on JS sourcemaps.

Workaround: `node --max-old-space-size=2000 ./node_modules/.bin/gulp {TASK}`

(2000 is 2GB)

- [ ] fix the underlying issue
- [ ] remove the heap config workaround from `scripts/publish/npm_publish.sh`

**Comment by robwormald** [comment]
i merged this one yesterday to try mitigate the JS/source map end without getting too invasive, and it seemed to help somewhat, but might need a bit more work. https://github.com/angular/angular/pull/5208

See also https://github.com/angular/angular/issues/5223 as we're likely to be re-orging this

**Comment by pkozlowski-opensource** [comment]
I was _so_ happy that we can run this bundling in parallel :-) But yeh, if it gives trouble to our build infrastructure we can easily re-shape it to do things sequentially. Provided that we are sure that this is the root cause, of course.

**Comment by IgorMinar** [comment]
it might be enough to just refactor the code bundle sequentially instead of using `Q.all`.

**Comment by robwormald** [comment]
I have a pending PR of the new system builder, that allows separation
between the tracing step and the writing to disk step. It definitely is
much faster and lighter weight. Coming soon.

**Comment by alfonso-presa** [comment]
Just in case it helps. I think this trace might be related to this issue. Running `gulp build` on OS X El Capitan with node v4.2.4 :

```
[12:38:10] Starting '!bundle.testing'...

<--- Last few GCs --->

  113851 ms: Mark-sweep 1380.6 (1456.9) -> 1383.2 (1456.9) MB, 1259.5 / 0 ms [allocation failure] [GC in old space requested].
  115110 ms: Mark-sweep 1383.2 (1456.9) -> 1383.2 (1456.9) MB, 1258.9 / 0 ms [allocation failure] [GC in old space requested].
  116382 ms: Mark-sweep 1383.2 (1456.9) -> 1378.1 (1456.9) MB, 1272.0 / 0 ms [last resort gc].
  117639 ms: Mark-sweep 1378.1 (1456.9) -> 1379.9 (1456.9) MB, 1256.3 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0xf6261cb4629 <JS Object>
    2: _serializeMappings(aka SourceMapGenerator_serializeMappings) [/Users/alfonso/Workspace/angular/node_modules/systemjs-builder/node_modules/source-map/lib/source-map/source-map-generator.js:~289] [pc=0x111d3ae2115e] (this=0x1b111ff80b59 <a SourceMapGenerator with map 0x21a7c3960b41>)
    3: toJSON(aka SourceMapGenerator_toJSON) [/Users/alfonso/Workspace/angular/node_modules/systemjs-builder...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6
```

It doesn't happen if I only run `gulp build.js`

**Comment by sb8244** [comment]
Contributing a failure for `gulp build.js`. On node v4.2.1 El Capitan 16GB RAM.

```
[01:06:42] Starting '!bundle.testing'...

<--- Last few GCs --->

  140814 ms: Scavenge 1374.7 (1455.7) -> 1374.7 (1455.7) MB, 3.5 / 0 ms (+ 1.7 ms in 1 steps since last GC) [allocation failure] [incremental marking delayin
g mark-sweep].
  142008 ms: Mark-sweep 1374.7 (1455.7) -> 1296.6 (1455.7) MB, 1194.1 / 0 ms (+ 3.1 ms in 2 steps since start of marking, biggest step 1.7 ms) [last resort g

c].
  143279 ms: Mark-sweep 1296.6 (1455.7) -> 1296.6 (1455.7) MB, 1271.4 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x3e0e73637399 <JS Object>
    2: _serializeMappings(aka SourceMapGenerator_serializeMappings) [/Users/stephenbussey/personal/angular/node_modules/systemjs-builder/node_modules/source-
map/lib/source-map/source-map-generator.js:~289] [pc=0x13217c8fd240] (this=0x2b5a4f0e72d1 <a SourceMapGenerator with map 0x37c172fd5261>)
    3: toJSON(aka SourceMapGenerator_toJSON) [/Users/stephenbussey/personal/angular/node_modules/syste...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
[1]    95243 abort      $(npm bin)/gulp build.js
```

**Comment by IgorMinar** [comment]
@pkozlowski-opensource is looking into this. I don't think that we'll be able to reduce the memory usage easily without impacting the total build time, so let's increase the heap size by default when calling `gulp build.js` (or more specifically the bundle task that causes the OOM).

This will require spawning a new process from within the gulp process but that should be a big deal. We just need to make sure that it's done a cross-platform compatible way.

**Comment by alexeagle** [comment]
Why not allocate more memory to the entire gulp process, rather than start a sub-process? Is it to avoid parts of the build outside of bundling that start to rely on more available heap?

One solution is to use npm run-scripts in the `package.json` so the memory settings are saved with the gulp command to run, eg
```
"scripts": {
  "build": "node --max-old-space-size=2048 ./node_modules/.bit/gulp build.js"
}
```

(I'm trying this now because I have the same issue on Circle, which already uses `npm run build` rather than shell scripts as the build step)

**Comment by IgorMinar** [comment]
We have way too many tasks that would need to be replicated as npm scripts.

If we replicate just some then it's going to be confusing.

Or how do you propose we do it? Just have a generic "npm gulp" task that
can be parameterized? That could work but seems a bit odd and nonergonomic.

**Comment by pkozlowski-opensource** [comment]
FYI I'm currently looking at our build / bundling to see if I can pin-point one task that is the major culprit here.

**Comment by IgorMinar** [comment]
Isn't the issue caused only by the bundling tasks?

**Comment by pkozlowski-opensource** [comment]
After spending more time tracking down what is going on here are my observations:
* `bundles.js` tasks is not a problem on its own - if you run `gulp bundles.js` always finishes without OOM
* testing various mix of tasks I was systematically getting OOM with a combination of 'build.js.dev', 'build.js.prod', 'build.js.cjs' and web-worker bundles
* removing either 'build.js.cjs' or web-worker bundles source maps from the mix results in no OOM

Based on the above I tend to think that no bundling tasks is a problem on its own, but it is a combination of previous build tasks and source maping for web worker tasks make the whole thing "memory hungry". We can't exclude that build tasks are leaking memory and this is why we are facing those OOM issues.

I'm preparing a PR that re-organizes tasks to mitigate those problems.

**Comment by zackhall** [comment]
I am also hitting this issue with `gulp build` and `gulp build.js`. Is there a mitigation I can use to get around this?

**Comment by alexeagle** [comment]
workaround is `node --max_old_space_size=4000 ./node_modules/.bin/gulp ...`

**Comment by zackhall** [comment]
Thanks, @alexeagle. That worked for me!

**Comment by jocull** [comment]
Documentation should just cite `npm run build` as that takes care of it for you.

**Comment by pkozlowski-opensource** [comment]
This one is obsolete with all the recent build changes.

**Comment by jepetko** [comment]
on windows environment I needed:
```
"scripts": {
  "build": "node --max-old-space-size=2048 node_modules/gulp/bin/gulp.js"
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
}
```

run as usual:
```
npm run build
```

**Comment by Sukh9212** [comment]
folks just a question

I've been running into process out of memory issues while running ngc.
ngc runs fine for the first time when things are in JIT mode in main.ts, after first run, I change main.ts to AoT mode.
Then ngc takes a while to compile and the process ends up out of memory.

Is there a way to increase `--max-old-space-size` while running ngc?

**Comment by IgorMinar** [comment]
Please file a new issue as this is unrelated to the old issue.

**Comment by Sukh9212** [comment]
there is issue #12184 which is also closed.
I couldn't find a solution anywhere to this

**Comment by AjayShelar** [comment]
https://www.npmjs.com/package/increase-memory-limit


npm install -g increase-memory-limit
Run from the root location of your project:
run inside project == increase-memory-limit
worked for me.
